### DISCUSSION TITLE
If factory function returns null, does GetOrSet save null in cache?

### DISCUSSION THREAD

**Discussion by pbolduc** [ZiggyCreatures/FusionCache#207] [Q&A] [answered]
I have a situation where my factory method will return null of a record is not found. Will the GetOrSet  / GetOrSetAsync store null in the cache, or does it just return null and nothing is cache?  I do not want null to be cached.  It wasn't clear from the docs (or I missed it).

**Comment by jodydonetti** [comment, marked as answer]
Hi @pbolduc and thanks for using FusionCache!

FusionCache caches `null` like any other value, because normally not caching `null` will lead to problems.

For example let's say that someone asks your service for `/product/123` and that product doesn't exist: the factory goes to the database, returns `null`, but `null` is not cached. This means that evey single request for it will go to the database, again and again.

**Basically that is a DoS/DDoS waiting to happen.**

As an example, this is what [StackOverflow does](https://nickcraver.com/blog/2019/08/06/stack-overflow-how-we-do-app-caching/):

> It also allows the caching of a null value and makes that handling not special-cased. In other words, you can tell the difference between _“It’s not cached”_ and _“We looked it up. It was null. We cached that null. STOP ASKING!”_

Having said that, if you think this is the good thing to do in your scenario and you really need to not cache `null`s (or, I would suggest, at least to cache them for a lower `Duration`), you can use [Adaptive Caching](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/AdaptiveCaching.md) to achieve that.

Hope this helps.

**Comment by jodydonetti** [comment]
> It wasn't clear from the docs (or I missed it)

It's in the main readme, in the "Something more" details:

![image](https://github.com/ZiggyCreatures/FusionCache/assets/1010086/13e39eba-52cc-48f7-8cf6-16e21909d063)

Good catch though: probably I should put it also somewhere else to be more prominent.

**Reply by pbolduc** [reply]
You are awesome!

**Reply by jodydonetti** [reply]
Thanks @pbolduc 😊 !

If you'd like, let me know which path you took (cache for less or don't cache).

Also, can you please mark one of my comment as the answer?

Thanks!

**Reply by pbolduc** [reply]
I used the adaptive caching approach. In my case, I realized after the service I was calling would return empty collection if the data was not found. I cache it for 5 minutes, otherwise I cache for 1 day - https://github.com/bcgov/jag-traffic-courts-online/blob/d7f6ed4a6b4a09301d82a6383fdfeaa963b7fecc/src/backend/TrafficCourts/Citizen.Service/Services/Tickets/Search/TicketSearchService.cs#L198-L226

**Reply by jodydonetti** [reply]
Got it, thanks for sharing!
